io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java

```diff
@@ -0,0 +1,477 @@
+	 * @param netEssCapacity net capacity of the battery.
+	 * @return the amount of energy that is limited.
+	 */
+	private int getLimitEnergy(int netEssCapacity) {
+
+		// Usable capacity based on minimum SoC from Limit total discharge and emergency
+		// reserve controllers.
+		var limitSoc = 0;
+		for (ControllerEssLimitTotalDischarge ctrl : this.ctrlLimitTotalDischarges) {
+			limitSoc = TypeUtils.max(limitSoc, ctrl.getMinSoc().get());
+		}
+		for (ControllerEssEmergencyCapacityReserve ctrl : this.ctrlEmergencyCapacityReserves) {
+			limitSoc = TypeUtils.max(limitSoc, ctrl.getActualReserveSoc().get());
+		}
+		this.channel(TimeOfUseTariffController.ChannelId.MIN_SOC).setNextValue(limitSoc);
+
+		return netEssCapacity /* [Wh] */ / 100 * limitSoc;
+	}
+
+	/**
+	 * Apply the mode OFF logic.
+	 */
+	private void modeOff() {
+		// Update the timer.
+		this._setCharged(false);
+		this._setDelayed(false);
+		this._setChargeValue(null);
+		this._setStateMachine(StateMachine.STANDBY);
+	}
+
+	/**
+	 * This is only to visualize data for better debugging.
+	 */
+	private void updateVisualizationChannels() {
+
+		if (this.schedule == null || this.schedule.periods.isEmpty()) {
+			// Values are not yet calculated.
+			return;
+		}
+
+		// Update the timer.
+		this.calculateChargedTime.update(this.getChargedChannel().getNextValue().orElse(false));
+		this.calculateDelayedTime.update(this.getDelayedChannel().getNextValue().orElse(false));
+
+		// First period is always the current period.
+		final var period = this.schedule.periods.get(0);
+
+		// Set the channels
+		this._setQuarterlyPrices(period.price);
+		this._setPredictedConsumption(period.consumptionPrediction);
+		this._setPredictedProduction(period.productionPrediction);
+		this._setGridEnergyChannel(period.gridEnergy);
+		this._setChargeDischargeEnergyChannel(period.chargeDischargeEnergy);
+	}
+
+	@Override
+	public Timedata getTimedata() {
+		return this.timedata;
+	}
+
+	@Override
+	public CompletableFuture<? extends JsonrpcResponseSuccess> handleJsonrpcRequest(User user, JsonrpcRequest request)
+			throws OpenemsNamedException {
+
+		user.assertRoleIsAtLeast("handleJsonrpcRequest", Role.GUEST);
+
+		switch (request.getMethod()) {
+
+		case GetScheduleRequest.METHOD:
+			return this.handleGetScheduleRequest(user, GetScheduleRequest.from(request));
+
+		default:
+			throw OpenemsError.JSONRPC_UNHANDLED_METHOD.exception(request.getMethod());
+		}
+	}
+
+	/**
+	 * Handles a GetScheduleRequest.
+	 *
+	 * @param user    the User0
+	 * @param request the GetScheduleRequest
+	 * @return the Future JSON-RPC Response
+	 */
+	private CompletableFuture<? extends JsonrpcResponseSuccess> handleGetScheduleRequest(User user,
+			GetScheduleRequest request) {
+
+		var now = TimeOfUseTariffUtils.getNowRoundedDownToMinutes(this.componentManager.getClock(), MINUTES_PER_PERIOD);
+		var fromDate = now.minusHours(3);
+
+		var channeladdressPrices = new ChannelAddress("ctrlEssTimeOfUseTariff0", "QuarterlyPrices");
```

Comment:
@venu-sagar Why don't you use `this.id()` here, instead of hard-coded `ctrlEssTimeOfUseTariff0`. Won't this break if somebody configures a different ID?